rfcs/proposed/task_group_dynamic_dependencies/README.md

```diff
@@ -0,0 +1,366 @@
+
+    void add_another_task(tbb::task_group& tg, int work_id) {
+        tbb::task_handle new_task = tg.defer([=] { do_work(work_id); });
+
+        for (tbb::task_handle& p : users::find_predecessors(work_id)) {
+            new_task.add_predecessor(p);
+        }
+
+        tg.run(std::move(new_task));
+    }
+
+While the graph, as shown below, is simple, the completion status of the predecessors
+is unknown. Therefore, for ease of use, `task_handle` should be usable as a dependency
+regardless of state of the task it represents. Any predecessor that is already
+completed when it is added as a predecessor will not delay the start of the dependent
+task.
+
+<img src="unknown_states.png" width=400>
+
+#### Recursive Decomposition
+
+This example is a version of merge-sort (with many of the details left out). 
+Assume an initial task executes the function shown below as its body, and the
+function implements that task, and also serves as the body for the recursively
+decomposed pieces. The range of the sequence is defined by `b` (beginning) and
+`e` (end). Most of the implementation details of merge sort are abstracted
+into the following helper functions: `users::do_serial_sort`, `users::create_left_range`,
+`users::create_right_range`, and `users::do_merge`.
+
+    template<typename T>
+    void merge_sort(tbb::task_group& tg, T b, T e) {
+        if (users::range_is_too_small(b, e)) {
+            // base-case when range is small
+            users::do_serial_sort(b, e);
+        } else {
+            // calculate left and right ranges
+            T lb, le, rb, re;
+            users::create_left_range(lb, le, b, e);
+            users::create_right_range(rb, re, b, e);
+
+            // create the three tasks
+            tbb::task_handle sortleft =
+                tg.defer([lb, le, &tg] {
+                    merge_sort(tg, lb, le);
+                });
+            tbb::task_handle sortright =
+                tg.defer([rb, re, &tg] {
+                    merge_sort(tg, rb, re);
+                });
+            tbb::task_handle merge =
+            tg.defer([lb, le, rb, re, &tg] {
+                    users::do_merge(tg, lb, le, rb, re);
+                });
+
+            // add predecessors for new merge task
+            merge.add_predecessor(sortleft);
+            merge.add_predecessor(sortright);
+
+            // insert new subgraph between currently executing
+            // task and its successors
+            tbb::task_group::transfer_successors_to(merge);
+
+            tg.run(std::move(sortleft));
+            tg.run(std::move(sortright));
+            tg.run(std::move(merge));
+        }
+    }
+
+This task tree matches the one shown earlier for merge-sort. 
+
+## Open Questions in Design
+
+Questions that should be addressed before moving into `experimental` and providing a preview feature:
+- Are the suggested APIs sufficient?
+- Are there additional use cases that should be considered that we missed in our analysis?
```

Comment:
The APIs seem sufficient for the considered use cases, except for semantical changes that need to be done. Some additional use cases to analyze were proposed, and currently are targeted as "post-experimental" work. I think the first question can be closed, and the second one can be either also closed or moved to that later stage.

Reply:
Applied